docs/modules/ROOT/pages/adr/0030-function-revisions.adoc

```diff
@@ -0,0 +1,123 @@
+
+== DIY Function Revisions
+It is possible to leverage the current Compositon Revisions together with Composition Functions, however some changes are necessary.
+
+We have to make sure ourselves that the right functions are deployed. For example the most current version plus the last 4 versions. Also, they have to be deployed with distinct name, so that they can be referenced directly in the composition.
+
+.Example: Function with a version suffix for the name
+[source, yaml]
+----
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  annotations:
+    argocd.argoproj.io/sync-options: SkipDryRunOnMissingResource=true
+    argocd.argoproj.io/sync-wave: '-40'
+  name: function-appcat-v4-120-2 <1>
+spec:
+  package: ghcr.io/vshn/appcat:v4.120.2-func <1>
+  runtimeConfigRef:
+    name: function-appcat
+  skipDependencyResolution: true
+----
+<1> The numbers in the package tag and the name of the function must match
+
+Then it can directly be referenced in the composition:
+
+[source, yaml]
+----
+apiVersion: apiextensions.crossplane.io/v1
+kind: Composition
+metadata:
+  name: vshnkeycloak.vshn.appcat.vshn.io
+  labels:
+    metadata.appcat.vshn.io/revision: v4.120.2 <2>
+spec:
+  compositeTypeRef:
+    apiVersion: vshn.appcat.vshn.io/v1
+    kind: XVSHNKeycloak
+  mode: Pipeline
+  pipeline:
+    - functionRef:
+        name: function-appcat-v4-120-2 <1>
+      input:
+----
+<1> Reference to the function
+<2> Label that can be used to reference a Composition Revision in a claim or composite via a selector
+
+Once `function-appcat:v4.120.3-func` gets released and the composition changes, then there will still be the old revision that explicitly references the old function. Each claim and composite that point to that revision will reconcile against the old function, not against the new one.
+
+This emulates the same behavior as standard Composition Revisions. If at any point in the future Crossplane implements a proper support for function revision, our own solution should be sunset and instead be switched to Crossplanes solution.
+
+== High level flow
+The following flowchart shows how some additional logic in the maintenance logic for each service can enable a transition from one revision to the next.
+
+This is independent on whether we use our DIY function revisions
+
+[mermaid,flow,png]
+....
+flowchart TD
+    start[Claim/Composite created] --> test{Test OpenShift Cluster?}
+    test --> |Yes| testcluster[Always use latest revision]
+    testcluster --> stop
+    test --> |No| prod[Set revision policy to manual]
+    prod --> maint[Maintenance Cronjob triggers]
+    maint --> first{Is first maintenance for instance?}
+    first --> |Yes| autpolicy[Set revision policy to auto]
+    first --> |No| selector[Set revision selector to most current revision]
```

Comment:
I don't understand this logic when it's NO, it has to go back to policy manual and remain on the older version until its maintenance is triggered and then it can go to the happy path yes and set to auto no?

Reply:
No, if it's not the first maintenance, then it's already set to automatic, so we can skip settings the policy. We only have to change from manual to automatic on the first maintenance, when we set the selector for the revision.

Just for the record, here's the rendered chart:
![image](https://github.com/user-attachments/assets/2d9113a8-8ae5-4379-aa3a-d0e05294e1b0)